File: src/scripts/slidingWindow.lua
-- Sliding Window Counter Algorithm (Atomic Lua Script)
-- KEYS[1] = sorted set key
-- ARGV[1] = limit (max requests in window)
-- ARGV[2] = windowMs (window size in milliseconds)
-- ARGV[3] = now (current epoch milliseconds)
--
-- Returns: {allowed (0/1), remaining, limit}

local key = KEYS[1]
local limit = tonumber(ARGV[1])
local windowMs = tonumber(ARGV[2])
local now = tonumber(ARGV[3])

-- Calculate window start
local windowStart = now - windowMs

-- Remove expired entries outside the window
redis.call('ZREMRANGEBYSCORE', key, 0, windowStart)

-- Count current requests in the window
local count = redis.call('ZCARD', key)

if count < limit then
    -- Allowed: add this request's timestamp
    -- Use now as both score and member; append random suffix to handle same-ms requests
    local member = now .. ':' .. math.random(1000000)
    redis.call('ZADD', key, now, member)
    redis.call('PEXPIRE', key, windowMs + 1000)
    return {1, limit - count - 1, limit}
else
    -- Rejected: over the limit
    redis.call('PEXPIRE', key, windowMs + 1000)
    return {0, 0, limit}
end

File: src/core/rateLimiter/SlidingWindow.ts
import { RateLimiter } from './RateLimiter';
import { Algorithm, RateLimitConfig, RateLimitResult } from './types';
import { evalScript } from '../../redis/luaLoader';
import { slidingWindowKey } from '../../redis/keyBuilder';
import { getRedisClient } from '../../redis/client';

/**
 * Sliding Window Counter rate limiter.
 *
 * Tracks exact request timestamps in a sorted set.
 * More accurate than Fixed Window — avoids boundary bursts.
 */
export class SlidingWindow extends RateLimiter {
  constructor(config: RateLimitConfig) {
    super(config);
    if (!this.config.windowSeconds) {
      this.config.windowSeconds = 60;
    }
  }

  async check(clientId: string, route: string): Promise<RateLimitResult> {
    const key = slidingWindowKey(clientId, route);
    const limit = this.config.limit;
    const windowMs = (this.config.windowSeconds ?? 60) * 1000;
    const now = Date.now();

    const result = (await evalScript('slidingWindow', [key], [
      limit,
      windowMs,
      now,
    ])) as number[];

    const allowed = result[0] === 1;
    const remaining = result[1];
    const resetAt = now + windowMs;

    const retryAfter = allowed ? undefined : Math.ceil(windowMs / 1000);

    return {
      allowed,
      remaining,
      limit,
      resetAt,
      retryAfter,
      algorithm: Algorithm.SLIDING_WINDOW,
      clientId,
      route,
    };
  }

  async reset(clientId: string, route: string): Promise<void> {
    const key = slidingWindowKey(clientId, route);
    const redis = getRedisClient();
    await redis.del(key);
  }

  async getQuota(clientId: string, route: string): Promise<Record<string, unknown>> {
    const key = slidingWindowKey(clientId, route);
    const redis = getRedisClient();
    const windowMs = (this.config.windowSeconds ?? 60) * 1000;
    const now = Date.now();
    const windowStart = now - windowMs;

    // Remove expired entries and count
    await redis.zremrangebyscore(key, 0, windowStart);
    const requestsInWindow = await redis.zcard(key);

    return {
      requestsInWindow,
      limit: this.config.limit,
      windowSeconds: this.config.windowSeconds ?? 60,
      windowStart,
    };
  }
}
